src/frontend/src/components/toast.ts

```diff
@@ -1,60 +1,67 @@
 import { TemplateElement } from "../utils/lit-html";
-import { irregularity } from "./irregularity";
+import { Chan } from "../utils/utils";
+import { closeIcon, warningIcon } from "./icons";
 
-export const removeToast = (toast: HTMLElement): void => {
-  toast.addEventListener("animationend", () => {
-    toast.remove();
-  });
-  toast.classList.add("c-toast--closing");
-};
-
-const createToast = (message: TemplateElement): HTMLDivElement => {
-  const toastEl = document.createElement("div");
-  toastEl.classList.add("c-toast");
-  render(
-    irregularity({
-      message,
-      closeFn: () => removeToast(toastEl),
-    }),
-    toastEl
-  );
-  return toastEl;
-};
-
-/**
- * Get the container for the toasts, or create it if it doesn't exist.
- * @returns the container reference for the toasts
- */
-const getOrCreateToaster = (): Element | null => {
-  const toasterEl = document.querySelector("[data-toaster]");
-  if (!toasterEl) {
-    // generic toaster that will be appended to the body
-    const toasterEl = document.createElement("div");
-    toasterEl.classList.add("c-toasts");
-    // container for the toasts. Two reasons for this:
-    // 1. we can reuse the l-container class for styling
-    // 2. the toasts will only be as wide as the container (user can still interact with the rest of the page)
-    const containerEl = document.createElement("div");
-    containerEl.classList.add("l-container");
-    // Attribute used to retrieve the element to append toasts to
-    containerEl.setAttribute("data-toasts", "");
-    toasterEl.setAttribute("data-toaster", "");
-    toasterEl.appendChild(containerEl);
-    document.body.appendChild(toasterEl);
-    return containerEl;
-  }
-  return toasterEl.querySelector("[data-toasts]");
-};
+// A toast element containing a message, and the (static) list of all toasts. The `Toast` element
+// must be an object because we (lit) use object-references in `repeat` to figure out which DOM elements
+// to replace.
+type Toast = { message: TemplateElement };
+const toasts: Toast[] = []; // the "toasts" singleton
 
 /**
  * A lot like console.error, but with a toast.
  * use like: toast.error("Something went wrong");
  */
 export const toast = {
-  error: (message: TemplateElement): void => {
-    const toasterEl = getOrCreateToaster();
-    const toastEl = createToast(message);
-    toasterEl?.appendChild(toastEl);
+  error: (message: Toast["message"]): void => {
+    toasts.push({ message });
+    renderToasts();
   },
 };
+
+// Render or rerender toasts.
+const renderToasts = () => {
+  render(
+    html` <div class="c-toasts l-container">
+      ${repeat(toasts, (x) => x, toastTemplate)}
+    </div>`,
+    document.body
+  );
+};
+
+const toastTemplate = (toast: Toast): TemplateResult => {
+  const closing = new Chan<boolean>(false);
+  const closeToast = () => closing.send(true);
+
+  const removeToast = () => {
+    const toastIndex = toasts.indexOf(toast);
```

Comment:
does that work if user closes the toast in the middle?

```
toast 1
toast 2 <--- user close this toast
toast 3
```

wouldn't it be safer to set id (`symbol`) to the toast and closes the particular one?

Reply:
> does that work if user closes the toast in the middle?

Definitely, any particular reason you think it might not work in all cases?

> wouldn't it be safer to set id (symbol) to the toast and closes the particular one?

I think the current way is fine as long as the `Toast` is an object (see comment near the type) because then this will be compared by reference, which are just as unique as `Symobl`s (if I understand the question correctly)

Reply:
> The indexOf() method compares searchElement to elements of the array using [strict equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) (the same algorithm used by the === operator).
> (about ===) If both operands are objects, return true only if they refer to the same object.


https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality

Reply:
I've trust issue in comparing object, too much Java probably. That's why I generally rather like to test on a unique identifier.

Good then. Thanks for the feedback.